### DISCUSSION TITLE
Does usage of dynamic schema cause performance issues?

### DISCUSSION THREAD

**Discussion by dykono** [milvus-io/milvus#30009] [Q&A and General discussion]
I am considering migrating data from one collection to a new one configure with dynmic schema. However, I am concerned that I can have performance issues after that.

**Comment by yhmo** [comment]
In fact, dynamic schema is a JSON field. 
- Once you enable dynamic schema, milvus creates a hidden JSON field for the collection. 
- When you insert data row by row(each row is a dict of key-value format), for the keys that doesn't match any field names, mivus will put the key and its value into the JSON field.
- If you search with an expression, the expression contains key that doesn't match any field names, milvus will look up the key in the dynamic JSON field.


The performance of filtering on a JSON field is not good. If a key exists in every row, then I recommend you define a field for the key, because filtering on a VARCHAR/INT/FLOAT/BOOL field(with index) is better than filtering on a JSON field.

**Reply by dykono** [reply]
Got it! Thanks for the explanation.

**Comment by xiaofan-luan** [comment]
please wait for our implementation on indexing JSON field.
yes, filtering on dynamic schema has performance regression

**Reply by dykono** [reply]
Thank you!
